src/libs/safe-path/src/scoped_dir_builder.rs

```diff
@@ -0,0 +1,249 @@
+
+        // root directory doesn't exist
+        ScopedDirBuilder::new(rootfs_path.join("__does_not_exist__")).unwrap_err();
+
+        // root is a file
+        fs::write(rootfs_path.join("txt"), "test").unwrap();
+        ScopedDirBuilder::new(rootfs_path.join("txt")).unwrap_err();
+
+        let mut builder = ScopedDirBuilder::new(&rootfs_path).unwrap();
+
+        // file with the same name already exists.
+        builder
+            .create_with_absolute_path(rootfs_path.join("txt"))
+            .unwrap_err();
+        // parent is a file
+        builder.create("/txt/a").unwrap_err();
+        // Not starting with root
+        builder.create_with_absolute_path("/txt/a").unwrap_err();
+        // creating "." without recursive mode should fail
+        builder
+            .create_with_absolute_path(rootfs_path.join("."))
+            .unwrap_err();
+        // parent doesn't exist
+        builder
+            .create_with_absolute_path(rootfs_path.join("a/b"))
+            .unwrap_err();
+        builder.create("a/b/c").unwrap_err();
+
+        let path = builder.create("a").unwrap();
+        assert!(rootfs_path.join("a").is_dir());
+        assert_eq!(path.target(), rootfs_path.join("a").canonicalize().unwrap());
+
+        // Creating an existing directory without recursive mode should fail.
+        builder
+            .create_with_absolute_path(rootfs_path.join("a"))
+            .unwrap_err();
+
+        // Creating an existing directory with recursive mode should succeed.
+        builder.recursive(true);
+        let path = builder
+            .create_with_absolute_path(rootfs_path.join("a"))
+            .unwrap();
+        assert_eq!(path.target(), rootfs_path.join("a").canonicalize().unwrap());
+        let path = builder.create(".").unwrap();
+        assert_eq!(path.target(), rootfs_path.canonicalize().unwrap());
+
+        builder.mode(0o740);
+        let path = builder.create("a/b/c/d").unwrap();
+        assert_eq!(
+            path.target(),
+            rootfs_path.join("a/b/c/d").canonicalize().unwrap()
+        );
+        assert!(rootfs_path.join("a/b/c/d").is_dir());
+        assert_eq!(
+            rootfs_path.join("a").metadata().unwrap().mode() & 0o777,
+            DIRECTORY_MODE_DEFAULT,
+        );
+        assert_eq!(
+            rootfs_path.join("a/b").metadata().unwrap().mode() & 0o777,
+            0o740
+        );
+        assert_eq!(
+            rootfs_path.join("a/b/c").metadata().unwrap().mode() & 0o777,
+            0o740
+        );
+        assert_eq!(
+            rootfs_path.join("a/b/c/d").metadata().unwrap().mode() & 0o777,
+            0o740
+        );
+
+        // Creating should fail if some components are not directory.
+        builder.create("txt/e/f").unwrap_err();
+        fs::write(rootfs_path.join("a/b/txt"), "test").unwrap();
+        builder.create("a/b/txt/h/i").unwrap_err();
+    }
+
+    #[test]
+    fn test_create_root() {
+        // root directory doesn't exist
```

Comment:
Uh.. that seems incorrect, I'm not sure what you're getting at with this comment.